lib/fizzy/stack.hpp

```diff
@@ -115,7 +115,7 @@ class OperandStack
         m_top = m_bottom - 1;
 
         const auto local_variables = std::copy_n(args, num_args, m_locals);
-        std::fill_n(local_variables, num_local_variables, 0);
+        std::fill_n(local_variables, num_local_variables, Value{});
```

Comment:
Hm, I'm confused by union initialization again. Does default initialization like this set the active member of the union?

Reply:
I'm confused too. This was only found by GCC coverage build.

The `0` is definitely wrong because it calls the special constructor and inits `.i32` only.

But the default constructor should be wrong too because it should only init the first field and `.i32` is the first one. This works however for current compilers. I wander if MSan can help here. Would be good if you can check the spec yourself.

One solution to solve this is to provide own implementation of the default constructor which inits `.i64` (or define `i64 = 0;`). But with that the `Value` looses some properties like "is trivially constructable".

Second solution is to move `.i64` to the first position. This guarantees the `Value{}` to be fully zero-initialized. Same for `FizzyValue` in C API.

But I don't see any sanity solution to init locals with zero without knowing the type.

Reply:
I don't believe I'm doing it again to myself, but:
1. The `Value{}` is value-initialization.
2. if T is a class type with a default constructor that is neither user-provided nor deleted (that is, it may be a class with an implicitly-defined or defaulted default constructor), the object is zero-initialized.
3. Zero-initialization: If T is a union type, the first non-static named data member is zero-initialized and all padding is initialized to zero bits.

So I think the `Value{}` is correct here as it zeros all object's bytes.

Reply:
It seems to depend on the interpretation of what is "padding" for a union - whether unused bits of unactive members are unions. I see some people having doubts about this here: https://stackoverflow.com/questions/11812555/how-to-zero-initialize-an-union#comment15704643_11812606 I have not found a definite answer to this yet.

However, I had a concern about something else (perhaps not directly related to this PR): we zero-initialize `Value`s of locals here, which might or might not choose the active member of the union (I have not found the answer to this either). Then function may access them right away with `local.get`, then some other operation may access some member of it - can be any type member. Then this may or may not be UB.

We can add a test for this, but as I remember sanitizers didn't catch union type punning.

And if we're required to set the right type, then there's no way around needing the types of locals when initializing.

Reply:
> We can add a test for this, but as I remember sanitizers didn't catch union type punning.

Well, the tests you added here (`execute.local_init`) already check this actually. Because checking the result of a function reads some specific member of `Value` union.

Reply:
I did some experiments and it looks like GCC and MSVC inits the whole object while Clang only the first element: https://godbolt.org/z/cWscY8.

Reply:
This may be a good reason to make `i64` the first member. Are there any advantages for it to be the second other than aesthetical?